packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.kt

```diff
@@ -110,6 +174,188 @@ class FileDownloader(
+      asset,
+      request,
+      responseBody,
+      updatesDirectory
+    )
+
+    val requestedUpdateId = request.header(EXPO_REQUESTED_UPDATE_ID_HEADER)
+
+    return responseBody.use { body ->
+      applyHermesDiff(
+        baseFile = launchAssetContext.baseFile,
+        diffBody = body,
+        destination = destination,
+        expectedBase64URLEncodedSHA256Hash = expectedBase64URLEncodedSHA256Hash,
+        asset = asset,
+        requestedUpdateId = requestedUpdateId
+      )
+    }
+  }
+
+  private data class LaunchAssetContext(val baseFile: File)
+
+  private fun prepareAssetForDiff(
+    asset: AssetEntity,
+    request: Request,
+    responseBody: ResponseBody,
+    updatesDirectory: File
+  ): LaunchAssetContext {
+    if (!asset.isLaunchAsset) {
+      failWithClosedBody(responseBody, "Received Hermes diff for non-launch asset ${asset.key}")
+    }
+
+    val currentUpdateIdHeader = request.header(EXPO_CURRENT_UPDATE_ID_HEADER)
+      ?: failWithClosedBody(
+        responseBody,
+        "Cannot apply Hermes diff without $EXPO_CURRENT_UPDATE_ID_HEADER header"
+      )
+
+    val db = database
+      ?: failWithClosedBody(responseBody, "Cannot apply Hermes diff without database access")
+
+    val currentUpdateId = try {
+      UUID.fromString(currentUpdateIdHeader)
+    } catch (e: IllegalArgumentException) {
+      failWithClosedBody(
+        responseBody,
+        "Invalid $EXPO_CURRENT_UPDATE_ID_HEADER header: $currentUpdateIdHeader",
+        e
+      )
+    }
+
+    val launchAssetEntity = db.updateDao().loadLaunchAssetForUpdate(currentUpdateId)
+      ?: failWithClosedBody(
+        responseBody,
+        "Launch asset not found for current update $currentUpdateIdHeader"
+      )
+
+    val launchAssetRelativePath = launchAssetEntity.relativePath ?: failWithClosedBody(
+      responseBody,
+      "Launch asset for update $currentUpdateIdHeader is missing a relative path"
+    )
+
+    val baseFile = File(updatesDirectory, launchAssetRelativePath)
+    if (!baseFile.exists()) {
+      failWithClosedBody(responseBody, "Base asset $baseFile is missing; cannot apply Hermes diff")
+    }
+
+    val actualBaseHash = try {
+      val hashBytes = UpdatesUtils.sha256(baseFile)
+      Base64.encodeToString(hashBytes, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
+    } catch (_: Exception) {
+      null
+    }
+
+    val expectedBaseHash = launchAssetEntity.expectedHash
+    if (expectedBaseHash != null && actualBaseHash != null && expectedBaseHash != actualBaseHash) {
+      logger.warn(
+        "Asset hash mismatch for update $currentUpdateIdHeader; expected=$expectedBaseHash actual=$actualBaseHash",
+        UpdatesErrorCode.AssetsFailedToLoad,
+        currentUpdateIdHeader,
+        asset.key
+      )
+      failWithClosedBody(
+        responseBody,
+        "Asset hash mismatch for update $currentUpdateIdHeader; expected=$expectedBaseHash actual=$actualBaseHash"
+      )
+    }
```

Comment:
hmm, I would hope that we don't need to re-check this at time of patch application. if it no longer matches, that'd mean that it somehow got corrupted on disk since the last time we downloaded/verified it?

Reply:
Again, just being defensive. I can remove it if it's overkill